### PR TITLE
add cost tracking for each bank during replay

### PR DESCRIPTION
**Pull request by taozhu-chicago**
#### Problem
Leaders use `cost_tracker` to ensure each block produced are with block limits. Same cost tracking and checking against limits should be symmetrically performed during replay, blocks exceed limits should fail replay.

#### Summary of Changes
- calling `bank.cost_tracker.try_add(tx)` during replaying, return error if block exceeds limits
- add feature gate since this change breaks consensus 

Feature Gate Issue: #29595 

**NOTE**: Until this point, cost_model and cost_tracker are exclusively used by leaders during block packing, so we enjoyed the freedom to change how it works without worrying much about breaking consensus. With this PR, future changes to `cost_model.rs`, `cost_tracker.rs`, `block_cost_limits.rs` and maybe other models that would change how transaction cost measures against block limits could break consensus, therefore need to be feature gated.

### CODE REVIEW

File: ledger/src/blockstore_processor.rs [hunk 1]
```diff
@@ -337,34 +338,49 @@ fn execute_batches(
...
 
+    if bank
+        .feature_set
+        .is_active(&feature_set::apply_cost_tracker_during_replay::id())
+    {
+        let mut cost_tracker = bank.write_cost_tracker().unwrap();
+        for tx_cost in &tx_costs {
+            cost_tracker
+                .try_add(tx_cost)
+                .map_err(TransactionError::from)?;
```
Comment (taozhu-chicago) [inline ledger/src/blockstore_processor.rs:353]: The feature gated change here to apply exact same cost tracking as leader did when produced the block. Fail with TransactionError should any limits were exceeded. If all blocks were produced honestly, check here should never fail, unless leader and validators running different version of cost_model/cost_tracker/block_cost_limits as mentioned in PR description.

File: ledger/src/blockstore_processor.rs [hunk 2]
```diff
@@ -337,34 +338,49 @@ fn execute_batches(
...
-                    &sanitized_txs,
-                    slice_start,
-                    index,
-                    &transaction_indexes,
-                );
-                slice_start = next_index;
-                tx_batches.push(tx_batch);
-                batch_cost = 0;
-            }
-        });
```
Comment (taozhu-chicago) [inline ledger/src/blockstore_processor.rs:367]: this block change is only due to `tx_costs` has changed from `vec<u64>` to `vec<TransactionCost>`, no other logical changes here.

**Comment by taozhu-chicago** [comment]
> > **NOTE**: Until this point, cost_model and cost_tracker are exclusively used by leaders during block packing, so we enjoyed the freedom to change how it works without worrying much about breaking consensus. With this PR, future changes to `cost_model.rs`, `cost_tracker.rs`, `block_cost_limits.rs` and maybe other models that would change how transaction cost measures against block limits could break consensus, therefore need to be feature gated.
> 
> To confirm,
> 
> * Prior to this change, the limits imposed by this files were only checked by leaders
> * But, the above wouldn't hypothetically prevent someone with mods from creating blocks that broke some rules
> * After this change, non-leaders will begin checking these rules in replay
> * Checking of these rules in replay could result in rejecting a block that previously would have been allowed through
> 
> And the last bullet is why this could affect consensus, and why we need a feature gate for this change and any subsequent modifications to these file. Is that all correct?

Exactly, all points correct. Checking during replay prevents large-than-limits blocks from being replayed, potentially causing validators skipping slots.

File: runtime/src/cost_tracker.rs [hunk 3]
```diff
@@ -29,6 +31,22 @@ pub enum CostTrackerError {
     WouldExceedAccountDataTotalLimit,
 }
 
+impl From<CostTrackerError> for TransactionError {
```
Comment (steviez) [inline runtime/src/cost_tracker.rs:34]: nit: Might be nice to have this in its' own PR to keep the feature-gate PR as slim as possible, but I guess we might run into unused code warning if we do that?

**Review by steviez — COMMENTED** [review]
Things look good to me, but looks like there is a merge issue that will require resolution and thus me looking again since a ship it would get dismissed.

File: ledger/src/blockstore_processor.rs [hunk 4]
```diff
@@ -337,34 +338,49 @@ fn execute_batches(
             let cost = tx_cost.sum();
             minimal_tx_cost = std::cmp::min(minimal_tx_cost, cost);
             total_cost = total_cost.saturating_add(cost);
-            cost
+            tx_cost
```
[the remark this answers is not in the document]
Comment (taozhu-chicago) [reply ledger/src/blockstore_processor.rs:341]: noted, can do it in a separate pr.

File: ledger/src/blockstore_processor.rs (hunk 1 above)
Comment (taozhu-chicago) [reply ledger/src/blockstore_processor.rs:353]: yeah, think so. Just a side note, removing enabled feature gate code is no small task, it should be treated at least as refactoring.

File: runtime/src/cost_tracker.rs (hunk 3 above)
Comment (taozhu-chicago) [reply runtime/src/cost_tracker.rs:34]: Like this idea, https://github.com/solana-labs/solana/pull/30267 for this, I will rebase once that's merged.

**Review by steviez — APPROVED** [review]
Nice, got the PR down to the bare minimum to introduce the feature gate!

**Comment by taozhu-chicago** [comment]
> Nice, got the PR down to the bare minimum to introduce the feature gate!

Thank you ❤️

**Pull request merged**